modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2220,29 +2223,35 @@ public CodegenOperation fromOperation(String path,
                                           Map<String, Schema> schemas,
                                           OpenAPI openAPI) {
         LOGGER.debug("fromOperation => operation: " + operation);
+        if (operation == null)
+            throw new RuntimeException("operation cannot be null in fromOperation");
+
         CodegenOperation op = CodegenModelFactory.newInstance(CodegenModelType.OPERATION);
         Set<String> imports = new HashSet<String>();
         if (operation.getExtensions() != null && !operation.getExtensions().isEmpty()) {
             op.vendorExtensions.putAll(operation.getExtensions());
-        }
 
-        if (operation == null)
-            throw new RuntimeException("operation cannot be null in fromOperation");
+            Object isCallbackRequest = op.vendorExtensions.remove("x-callback-request");
+            op.isCallbackRequest = Boolean.TRUE.equals(isCallbackRequest);
+        }
 
         // store the original operationId for plug-in
         op.operationIdOriginal = operation.getOperationId();
 
-        String operationId = getOrGenerateOperationId(operation, path, httpMethod);
-        // remove prefix in operationId
-        if (removeOperationIdPrefix) {
-            int offset = operationId.indexOf('_');
-            if (offset > -1) {
-                operationId = operationId.substring(offset + 1);
+        if (!op.isCallbackRequest) {
+            String operationId = getOrGenerateOperationId(operation, path, httpMethod);
+            // remove prefix in operationId
+            if (removeOperationIdPrefix) {
+                int offset = operationId.indexOf('_');
+                if (offset > -1) {
+                    operationId = operationId.substring(offset + 1);
+                }
             }
+            operationId = removeNonNameElementToCamelCase(operationId);
+            op.operationId = op.isCallbackRequest ? null : toOperationId(operationId);
```

Comment:
@devplayer0 may I know why the `operationId` is not set if callback is defined in the spec?

Reply:
@wing328 My initial thought was that since it's not a server side method, callbacks aren't really "operations" on an API per se. In the [OpenAPI example for callbacks](https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v3.0/callback-example.yaml) operation IDs are not used when defining the callback request.

If I were to not set the operation ID to null, I feel like it would be important to change the behaviour of generating a default operation ID if the user does not provide one - the runtime expression parts should probably be removed and the callback identifier (`onData` in the OpenAPI example) should probably be used as a prefix. What do you think?

Reply:
> In the OpenAPI example for callbacks operation IDs are not used when defining the callback request.

`operationId` (optional) is for uniquely identifying the operation:

```
Unique string used to identify the operation. The id MUST be unique among all operations described in the API. Tools and libraries MAY use the operationId to uniquely identify an operation, therefore, it is RECOMMENDED to follow common programming naming conventions.
```

In that example, there's only one operation so it's ok not to specify the operationId.

>  the runtime expression parts should probably be removed and the callback identifier (onData in the OpenAPI example) should probably be used as a prefix.

Perfectly fine with me that you want to change the default naming when the callback is specified. My experience told me that the default will not meet everyone's requirement so the users need to specify the "operationId" if they want a better `operationId`